---
title: "Haplotype recovery experiment"
author: "Zena Lapp"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, collapse = TRUE, comment = NA)
```

```{r}
# load libraries
library(tidyverse)
library(scales)
library(cowplot)
library(pROC)
library(ggupset)
library(broom.mixed)
library(ggpmisc)
library(ggpubr)
library(ggtext)
library(lme4)

# set plotting theme
theme_set(theme_bw() + 
            theme(text = element_text(size = 10), 
            strip.text.x = element_markdown(),
            strip.text.y = element_markdown(),
            axis.text.x = element_markdown(),
            axis.text.y = element_markdown(),
            strip.background = element_rect(fill="white",linetype='blank')))

# read in data
targ_overview <- read_csv('data/target_overview.csv') %>% 
  mutate(target = paste0('*', target, '*'))

ref_pairwise_dists <- read_csv('data/ref_pairwise_dists.csv') %>% 
  mutate(target = paste0('*', target, '*'))

metadat <- read_csv('data/metadata.csv') %>% 
  mutate(dens = factor(dens, levels = c('<1.5 g/µL', '1.5-75 g/µL', '≥75 g/µL')),
         target = paste0('*', target, '*'))

read_cts <- read_csv('data/read_counts.csv') %>% 
  mutate(dens = factor(dens, levels = c('<1.5 g/µL', '1.5-75 g/µL', '≥75 g/µL')),
         target = paste0('*', target, '*'))

dada2_long <- read_csv('data/dada2_long.csv') %>% 
  mutate(dens = factor(dens, levels = c('<1.5 g/µL', '1.5-75 g/µL', '≥75 g/µL')),
         target = paste0('*', target, '*'),
         hap_type = factor(hap_type, levels = c('Expected reference', 'Systematic error', 'Random error'))) 

missing_info <- read_csv('data/missing_info.csv') %>% 
  mutate(dens = factor(dens, levels = rev(c('<1.5 g/µL', '1.5-75 g/µL', '≥75 g/µL'))),
         target = paste0('*', target, '*'))

dada2_long_clinsamps <- read_csv('data/dada2_long_clinsamps.csv') %>% 
  mutate(target = paste0('*', target, '*'))

# set seed for reproducibility
set.seed(1)
```


### Marker overview

```{r}
targ_overview
```

### Mixtures

```{r}
strain_cols <- bind_cols(ref_strains = metadat %>% 
  separate_rows(ref_strains, ref_strain_pcts, sep = ';') %>% 
    filter(!is.na(ref_strains)) %>% 
  pull(ref_strains) %>% 
  unique(),
  strain_col = c("#256676", "#ffa500", "#a3d2a0", "#9c3b68", "#be8fb9", "#4bd6fd", "#8c0da4", "#3256c2", "#708e30", "#ef6fcc"))

strain_cols_leg <- adjustcolor(strain_cols$strain_col, alpha.f = 0.7)
names(strain_cols_leg) <- strain_cols$ref_strains

mixture_overview_dat <- metadat %>% 
  select(mix, ref_strains, ref_strain_pcts) %>% 
  unique() %>% 
  separate_rows(ref_strains, ref_strain_pcts, sep = ';') %>% 
  mutate(ref_strain_pcts = as.numeric(ref_strain_pcts),
         mix_strain = paste(mix, ref_strains)) %>%
  left_join(strain_cols) %>% 
  arrange(ref_strain_pcts) %>%
  mutate(mix_strain = factor(mix_strain, levels = unique(mix_strain)))

mixture_overview_plot <- mixture_overview_dat %>%
  ggplot(aes(x = factor(mix), y = ref_strain_pcts, fill = mix_strain)) +
  geom_col(fill = 'white') +
  geom_col(col = 'white') +
  scale_fill_manual(values = adjustcolor(mixture_overview_dat$strain_col, alpha.f = 0.7), guide = 'none') +
  labs(x = 'Mixture', y = 'Proportion of mixture', fill = 'Reference\nstrain') 

mixture_leg <- (mixture_overview_dat %>%
  ggplot(aes(x = factor(mix), y = ref_strain_pcts, col = ref_strains)) +
  geom_point() +
    scale_color_manual(values = strain_cols_leg, 
                    breaks = names(strain_cols_leg),
                    labels = paste0("<span style='color:",
                                    strain_cols_leg,
                                    "'>", 
                                    names(strain_cols_leg),
                                    "</span>")) +
    guides(color=guide_legend(ncol=1, override.aes = list(size=-1), byrow = TRUE)) +
    labs(x = 'Mixture', y = 'Proportion of mixture', 
         color = '       Reference\n       strain') +
  theme(legend.text = element_markdown(face = 'bold'),
        legend.spacing.y = unit(-0.1, 'cm')
        )) %>% 
  get_legend()

mixture_overview <- plot_grid(mixture_leg, NULL, mixture_overview_plot, 
                              nrow = 1, rel_widths = c(1, -0.3, 1))
mixture_overview
```

## Reference strain sequences

```{r}
strain_order <- rev(c('FUP', '7g8', '3D7', 'Tanzania', 'FCB', 'FCR3-Gambia', 'K1', 'HB3', 'V1/S', 'Dd2'))

(ref_strain_heatmaps <- ref_pairwise_dists %>% 
  filter(strain1 != 'FCR3-FMG' & strain2 != 'FCR3-FMG') %>% # not in any of mixtures
  mutate(strain1 = factor(strain1, levels = strain_order),
         strain2 = factor(strain2, levels = strain_order)) %>% 
  ggplot(aes(x = strain1, y = strain2, fill = pairwise_dist)) +
  facet_grid(~target) +
  geom_tile() +
  scale_fill_gradient(low = 'white', high = 'lightsteelblue3') +
  theme(axis.text.x = element_markdown(angle = 90, hjust = 1, vjust = 0.5, size = 7),
        axis.text.y = element_markdown(size = 7),
        strip.text = element_markdown()) +
  labs(x = element_blank(), y = element_blank(), fill = 'Pairwise SNV\ndistance') +
  coord_fixed())
```

Summary of pairwise distances

```{r}
# overall summary of pairwise distances
ref_pairwise_dists %>% 
  filter(strain1 != strain2) %>%
  pull(pairwise_dist) %>% 
  summary()
```

```{r, eval = FALSE}
# Selecting a second reference sequence - most distant from 3D7
ref_pairwise_dists %>% 
  filter(strain1 == '3D7') %>% 
  ggplot(aes(y = strain2, x = pairwise_dist)) +
  geom_boxplot() +
  geom_jitter(alpha = 0.5, height = 0.2)

ref_pairwise_dists %>% 
  filter(strain1 == '3D7' & strain2 == 'HB3') 
```


## Read summary

```{r}
n_samps_by_ct <- metadat %>% group_by(dens) %>% tally()

read_cts %>% 
  group_by(dens) %>% 
  summarize(raw_reads = sum(raw_reads),
         n_reads_dada2 = sum(n_reads_sample)) %>% 
  left_join(n_samps_by_ct) %>% 
  mutate(mean_all_raw = raw_reads/n,
            mean_all_dada2 = n_reads_dada2/n) 
```

Number of samples that returned usable reads

```{r}
dada2_long %>% 
  select(sample, target) %>% 
  n_distinct()
```

Per-sample read depth

```{r}
marker_cols <- c('#6929c4', '#1192e8', '#005d5d', 'lightsalmon4', '#9f1853', '#fa4d56')
names(marker_cols) <- unique(dada2_long$target)
marker_cols_light <- adjustcolor(marker_cols, alpha.f = 0.7)
names(marker_cols_light) <- names(marker_cols)

(read_by_sample <- read_cts %>% 
    mutate(dens = factor(gsub(' g/µL', '', dens), levels = c('<1.5', '1.5-75', '≥75'))) %>% 
  ggplot(aes(y = n_reads_sample, x = dens)) +
  geom_boxplot() +
  geom_point(aes(col = target), 
             position = position_jitterdodge(jitter.width = 0.8, jitter.height = 0), alpha = 0.3) +
  guides(color = guide_legend(show.legend = FALSE)) +
  scale_color_manual(values = marker_cols) +
    scale_y_continuous(labels = label_comma()) +
  facet_grid(~target, scales = 'free') +
  labs(x = 'Parasite density (genomes/µL)', y = 'Number of reads in sample\nthat passed all filtering') +
  theme(legend.position = 'none')) 
```

Number of haplotype occurrences per target

```{r}
(n_haps <- dada2_long %>% 
  ggplot(aes(x = target, fill = dens)) +
  geom_bar(col = 'white') +
  scale_fill_grey(start = 0.8, end = 0.4) +
  labs(x = 'Marker', y = 'Number of haplotype\noccurrences', fill = 'Parasite\ndensity'))
```

Median number of reads per target

```{r}
median(dada2_long$n_reads)
```



```{r}
mixture_overview <- plot_grid(mixture_leg, mixture_overview_plot, 
                              nrow = 1, rel_widths = c(1,  2))

mix_strain_overview <- plot_grid(mixture_overview,
                                 ref_strain_heatmaps, 
          nrow = 1, labels = 'AUTO', rel_widths = c(0.52, 1))

read_overview <- plot_grid(read_by_sample, n_haps, 
                           labels = c('C', 'D'), rel_widths = c(1, 0.5), 
                           align = 'h', axis = 'tb',
                           nrow = 1)

overview_fig <- plot_grid(mix_strain_overview, read_overview, ncol = 1)
ggsave('figures/overview.png', overview_fig, width = 10.5, height = 5)
```

## Amplicon sequences

### False positives

Overview numbers

```{r}
dada2_long %>% 
  group_by(hap_type) %>%
  summarize(n_type = n(),
            sum_reads = sum(n_reads)) %>% 
  ungroup() %>% 
  mutate(tot = sum(n_type),
         pct_count = round(n_type/sum(n_type)*100, 1),
         pct_reads = round(sum_reads/sum(sum_reads)*100, 1))
```
Sample-level overview

```{r}
mix_num <- dada2_long %>%
  mutate(mix = paste0(mix, as.numeric(dens))) %>% 
  pull(mix) %>% 
  unique()
mix_labs <- case_when(grepl('2', mix_num) ~ paste0('mix ', gsub('2', '', mix_num)))
names(mix_labs) <- mix_num
targs <- unique(unique(dada2_long$target))
names(targs) <- targs
densities <- levels(dada2_long$dens)
names(densities) <- densities

fp_rep_fig <- dada2_long %>%
  mutate(mix = paste0(mix, as.numeric(dens))) %>% 
  ungroup() %>%
    ggplot(aes(x = rep, y = n_reads, fill = hap_type)) +
    facet_grid(target ~ mix + dens, scales = 'free', space = 'free',
               labeller = as_labeller(c(mix_labs, targs, densities))) +
    geom_text(data = tibble(mix = c('A3', 'C3'), 
                          dens = factor('≥75 g/µL', levels = levels(dada2_long$dens)), 
                          target = '*ama1*',
                          hap_type = factor(NA, levels = levels(dada2_long$hap_type))),
          label = 'Missing', x = 2, y = 75000, col = 'grey50', size = 3, vjust = 0, hjust = 0.5) +
    geom_col(col = 'white', size = 0.1) +
  scale_fill_manual(values = (c('Expected reference' = 'grey70', 'Systematic error' = 'pink', 'Random error' = 'red')),
  breaks = levels(dada2_long$hap_type),
  labels = gsub(' ', '\n', levels(dada2_long$hap_type))) +
    scale_y_continuous(labels = label_comma()) +
    labs(x = 'Technical replicate', y = 'Read depth', fill = 'Haplotype\ncategory')

frf <- ggplot_gtable(ggplot_build(fp_rep_fig))
frf$widths[c(6,8,12,14,18,20,24,26,30,32,36,38)] <- 0.5*frf$widths[c(6,8,12,14,18,20,24,26,30,32)]
(fp_rep_fig <- as_ggplot(frf))
```

Proportion of reads and haplotype occurrences that are false positive

```{r}
fp_prop_dat <- dada2_long %>% 
  group_by(target, hap_type) %>%
  summarize(n_type = n(),
            n_reads = sum(n_reads)) %>%
  group_by(target) %>% 
  mutate(pct_type = n_type/sum(n_type),
         pct_reads = n_reads/sum(n_reads)) %>% 
  ungroup() %>% 
  select(-c(n_type, n_reads)) %>% 
  pivot_longer(c(pct_type, pct_reads)) %>% 
  mutate(name = ifelse(name == 'pct_type', 'Haplotypes', 'Reads')) 

(fp_prop_fig <- fp_prop_dat %>% 
  ggplot(aes(x = name, y = value, fill = hap_type)) +
    facet_grid(target~.) +
    geom_col() +
    scale_fill_manual(values = (c('Expected reference' = 'grey70', 
                                  'Systematic error' = 'pink',
                                  'Random error' = 'red')),
                      breaks = levels(dada2_long$hap_type),
  labels = gsub(' ', '\n', levels(dada2_long$hap_type))) +
    scale_y_continuous(breaks = pretty_breaks()) +
  labs(x = '', y = 'Proportion of total', fill = 'Haplotype category') + 
    coord_flip())
```

Summary statistics of false positives

```{r}
dada2_long %>% 
  group_by(target, hap_type) %>% 
  summarize(n_type = n(),
            n_reads = sum(n_reads)) %>%
  group_by(target) %>% 
  mutate(pct_type = round(n_type/sum(n_type)*100, 1),
         pct_reads = round(n_reads/sum(n_reads)*100, 1))

dada2_long %>% 
  group_by(dens, hap_type) %>% 
  summarize(n_type = n(),
            n_reads = sum(n_reads)) %>%
  group_by(dens) %>% 
  mutate(pct_type = round(n_type/sum(n_type)*100, 1),
         pct_reads = round(n_reads/sum(n_reads)*100, 1)) %>% 
  filter(hap_type == 'Random error')
```
Number of reads supporting true positives vs false positives

```{r}
dada2_long %>% 
  summarize(median_in = median(n_reads[hap_type == 'Expected reference']),
            median_not_in = median(n_reads[hap_type == 'Random error']),
            wilcox_p = wilcox.test(n_reads[hap_type == 'Expected reference'], n_reads[hap_type == 'Random error'])$p.value)
```

Read depth of false positives

```{r}
(fp_reads <- dada2_long %>%
  mutate(dens = factor(gsub(' g/µL', '', dens), levels = c('<1.5', '1.5-75', '≥75'))) %>% 
  ggplot(aes(x = dens, y = n_reads, col = hap_type)) +
  geom_jitter(alpha = 0.5, position = position_jitterdodge(dodge.width = 0.5, jitter.width = 0.5/3)) +
  # geom_hline(yintercept = 250, linetype = 'dotted') +
  scale_color_manual(values = c('grey70', 'pink', 'red')) +
  facet_grid(target~., scales = 'free', space = 'free') +
  scale_y_log10(labels = label_comma()) +
  labs(x = 'Parasite density (genomes/µL)', y = 'Number of reads', col = 'Haplotype category'))
```

Characteristics of false positives

```{r}
(fp_overview <- dada2_long %>% 
  filter(hap_type == 'Random error') %>% 
  left_join(targ_overview) %>% 
  mutate(same_len = ifelse(diff_3d7, 'Not expected length', 'Expected length')) %>% 
  group_by(target, dens, dada2_alias, same_len, hap_type) %>%
  summarize(dist_to_ref_1 = factor(case_when(same_len == 'Not expected length' ~ 'Not expected\nlength',
                                      min(dist_to_ref, na.rm = TRUE) == 1 ~ 'Hamming distance\nto 3D7 = 1',
                                      min(dist_to_ref, na.rm = TRUE) > 1 ~ 'Hamming distance\nto 3D7 > 1'),
                                   levels = c('Not expected\nlength', 'Hamming distance\nto 3D7 = 1', 
                                              'Hamming distance\nto 3D7 > 1')),
            count = n(),
            sum_reads = sum(n_reads)) %>% 
    distinct() %>% 
   mutate(dens = factor(gsub(' g/µL', '', dens), levels = c('<1.5', '1.5-75', '≥75'))) %>% 
  ggplot(aes(x = dens, y = count, fill = dist_to_ref_1)) +
  facet_grid(target~., 
             scales = 'free', space = 'free') +
  geom_col(binwidth = 1, fill = 'white') +
  geom_col(binwidth = 1, col = 'white') +
  scale_fill_grey(start = 0.7, end = 0.3) +
  labs(x = 'Parasite density (genomes/µL)', y = 'Number of haplotypes likely arising from random error',
       fill = '') +
  coord_flip())
```

```{r}
fp_plot <- plot_grid(fp_rep_fig,
                     plot_grid(fp_prop_fig +
                                 theme(legend.position = 'none'),
                               fp_overview,
                               fp_reads + 
                                 labs(y = 'Read depth') +
                                 theme(legend.position = 'none'),
                               NULL,
                               nrow = 1, rel_widths = c(1, 1.4, 1, 0.4), labels = c('B','C', 'D')),
                     ncol = 1, rel_heights = c(1.5, 1), labels = c('A',''))
ggsave(plot = fp_plot, filename = 'figures/fps.png', width = 13, height = 8)
```

### Identifying optimal thresholds 

```{r}
get_optims <- function(dat){
  cr <- roc(dat$in_samp, dat$var, levels = c('no', 'yes'), ci = TRUE)
  optims_all <- lapply(c(0.5, 1, 2), function(x){ # weight false positives and negatives differently
    optims <- coords(cr, "best", transpose = FALSE, best.method="youden",
                       best.weights = c(x, 0.5)) %>% 
      pivot_longer(everything(), names_to = 'metric', values_to = 'optim') %>% 
      mutate(cost = x)
    cis_thresh <- ci.coords(cr, x = 'best', best_method = 'youden', 
                            best.weights = c(x, 0.5),
                            best.policy = 'random') %>% 
      unlist() %>% 
      enframe() %>% 
      mutate(metric = gsub('[0-9]', '', name),
             level = gsub('[a-z]', '', name),
             level = case_when(level == 1 ~ 'ci025',
                               level == 2 ~ 'median',
                               level == 3 ~ 'ci975')) %>% 
      select(-name) %>% 
      pivot_wider(names_from = level, values_from = value) %>% 
      mutate(cost = x)
    full_join(optims, cis_thresh)
    }) %>% bind_rows()
  
  optims_sub_all <- optims_all %>% 
    rename(threshold = optim) %>% 
    mutate(cost = case_when(cost == 0.5 ~ 'FP 2x cost of FN',
                            cost == 1 ~ 'FP and FN equal cost',
                            cost == 2 ~ 'FN 2x cost of FP'),
           cost = factor(cost, levels = c('FN 2x cost of FP', 'FP and FN equal cost', 'FP 2x cost of FN')))
  
  return(list(roc_output = cr, optimal_thresholds = optims_sub_all))
}
```

```{r}
tp_read <- dada2_long %>% 
    filter(hap_type != 'Systematic error') %>% 
    mutate(in_samp = ifelse(hap_type == 'Random error', 'no', 'yes'),
           var = n_reads) %>% 
    select(in_samp, var) 
optim_depth <- get_optims(tp_read)

tp_prop <- dada2_long %>% 
    filter(hap_type != 'Systematic error') %>% 
    group_by(sample, target) %>% 
    mutate(in_samp = ifelse(hap_type == 'Random error', 'no', 'yes'),
           prop_samp = n_reads/sum(n_reads),
           var = prop_samp) %>% 
    ungroup() %>% 
    select(in_samp, var) 
optim_prop <- get_optims(tp_prop)

min_dist_1 <- dada2_long %>% 
  filter(min_dist == 1 & read_prop <= 1 & hap_type != 'Systematic error') %>% 
  mutate(in_samp = ifelse(hap_type == 'Random error', 'no', 'yes'),
         var = read_prop) 
optim_ratio <- get_optims(min_dist_1)
```

Sensitivity-specificity plots

```{r}
(depth_sens_spec <- 
  bind_rows(threshold = optim_depth$roc_output$thresholds,
            Sensitivity = optim_depth$roc_output$sensitivities,
            Specificity = optim_depth$roc_output$specificities) %>% 
  pivot_longer(c(Sensitivity, Specificity)) %>% 
  ggplot() +
    geom_vline(data = optim_depth$optimal_thresholds %>% 
                 filter(metric == 'threshold'),
             aes(xintercept = threshold), alpha = 0.75, size = 0.5, linetype = 'dotted') +
  geom_line(aes(x = threshold, y = value, linetype = name), size = 0.75) +
      geom_label(data = optim_depth$optimal_thresholds %>% 
                   filter(metric == 'threshold'),
             aes(x = threshold, label = round(threshold)),
             y = c(0.07, 0.15, 0.23), size = 3, fill = 'white', label.size = NA) +
  geom_point(data = optim_depth$optimal_thresholds %>% 
                   filter(metric == 'threshold'),
             aes(x = threshold, shape = cost), col = 'white', 
             y = c(0, 0.02, 0.04), size = 3) +
  geom_point(data = optim_depth$optimal_thresholds %>% 
                   filter(metric == 'threshold'),
             aes(x = threshold, shape = cost), 
             y = c(0, 0.02, 0.04), alpha = 0.75, size = 3) +
   geom_segment(data = optim_depth$optimal_thresholds %>% 
                   filter(metric == 'threshold'), 
                y = c(0, 0.02, 0.04), yend = c(0, 0.02, 0.04), 
                aes(x = ci025, xend = ci975)) +
  scale_color_manual(values = marker_cols) +
  scale_x_continuous(labels = label_comma(), breaks = pretty_breaks()) +
   coord_cartesian(xlim = c(0, 1500)) +
  labs(x = 'Read depth', y = 'Value', linetype = element_blank(), shape = '') +
  theme(panel.grid = element_blank(),
        legend.text = element_markdown()))

(prop_sens_spec <- 
  bind_rows(threshold = optim_prop$roc_output$thresholds,
            Sensitivity = optim_prop$roc_output$sensitivities,
            Specificity = optim_prop$roc_output$specificities) %>% 
  pivot_longer(c(Sensitivity, Specificity)) %>% 
  ggplot() +
    geom_vline(data = optim_prop$optimal_thresholds %>% 
                   filter(metric == 'threshold'),
             aes(xintercept = threshold), alpha = 0.75, size = 0.5, linetype = 'dotted') +
  geom_line(aes(x = threshold, y = value, linetype = name), size = 0.75) +
           geom_label(data = optim_prop$optimal_thresholds %>% 
                   filter(metric == 'threshold'),
             aes(x = threshold, label = round(threshold, 3)),
             y = c(0.07, 0.15, 0.23), size = 3, fill = 'white', label.size = NA) +
  geom_point(data = optim_prop$optimal_thresholds %>% 
                   filter(metric == 'threshold'),
             aes(x = threshold, shape = cost), col = 'white', y = c(0, 0.02, 0.04), size = 3) +
  geom_point(data = optim_prop$optimal_thresholds %>% 
                   filter(metric == 'threshold'),
             aes(x = threshold, shape = cost), y = c(0, 0.02, 0.04), alpha = 0.75, size = 3) +
   geom_segment(data = optim_prop$optimal_thresholds %>% 
                   filter(metric == 'threshold'), y = c(0, 0.02, 0.04), yend = c(0, 0.02, 0.04), aes(x = ci025, xend = ci975)) +
  scale_color_manual(values = marker_cols) +
  scale_x_log10(labels = label_comma()) +
  labs(x = 'Read proportion within a sample', y = 'Value', linetype = element_blank(), shape = '') +
  theme(panel.grid = element_blank(),
        legend.text = element_markdown()))

(ratio_sens_spec <- 
  bind_rows(threshold = optim_ratio$roc_output$thresholds,
            Sensitivity = optim_ratio$roc_output$sensitivities,
            Specificity = optim_ratio$roc_output$specificities) %>% 
  pivot_longer(c(Sensitivity, Specificity)) %>% 
  ggplot() +
    geom_vline(data = optim_ratio$optimal_thresholds %>% 
                   filter(metric == 'threshold'),
             aes(xintercept = threshold), alpha = 0.75, size = 0.5, linetype = 'dotted') +
  geom_line(aes(x = threshold, y = value, linetype = name), size = 0.75) +
           geom_label(data = optim_ratio$optimal_thresholds %>% 
                   filter(metric == 'threshold'),
             aes(x = threshold, label = round(threshold, 2)),
             y = c(0.07, 0.15, 0.23), size = 3, fill = 'white', label.size = NA) +
  geom_point(data = optim_ratio$optimal_thresholds %>% 
                   filter(metric == 'threshold'),
             aes(x = threshold, shape = cost), col = 'white', y = c(0, 0.02, 0.04),  size = 3) +
  geom_point(data = optim_ratio$optimal_thresholds %>% 
                   filter(metric == 'threshold'),
             aes(x = threshold, shape = cost), y = c(0, 0.02, 0.04), alpha = 0.75, size = 3) +
   geom_segment(data = optim_ratio$optimal_thresholds %>% 
                   filter(metric == 'threshold'), y = c(0, 0.02, 0.04), yend = c(0, 0.02, 0.04), aes(x = ci025, xend = ci975)) +
  scale_color_manual(values = marker_cols) +
  scale_x_log10(labels = label_comma()) +
  labs(x = 'Read ratio', y = 'Value', linetype = element_blank(), shape = '') +
  theme(panel.grid = element_blank(),
        legend.text = element_markdown()))
```

Optimal thresholds table

```{r}
bind_rows(optim_depth$optimal_thresholds %>% 
  filter(cost == 'FN 2x cost of FP') %>% 
  mutate(type = 'depth'),
optim_prop$optimal_thresholds %>% 
  filter(cost == 'FN 2x cost of FP') %>% 
  mutate(type = 'prop'),
optim_ratio$optimal_thresholds %>% 
  filter(cost == 'FN 2x cost of FP') %>% 
  mutate(type = 'ratio')) %>% 
  select(-cost)
```


```{r}
optim_dat <- dada2_long %>%
  group_by(sample, target) %>% 
  mutate(read_thresh = (optim_depth$optimal_thresholds %>% 
           filter(cost == 'FN 2x cost of FP' & metric == 'threshold') %>% 
           pull(threshold)),
         prop_thresh = (optim_prop$optimal_thresholds %>% 
           filter(cost == 'FN 2x cost of FP' & metric == 'threshold') %>% 
           pull(threshold)),
         ratio_thresh = (optim_ratio$optimal_thresholds %>% 
           filter(cost == 'FN 2x cost of FP' & metric == 'threshold') %>% 
           pull(threshold)),
         reads_lt_thresh = n_reads < read_thresh,
         prop_lt_thresh = n_reads/sum(n_reads) < prop_thresh,
         ratio_lt_thresh = read_prop <= ratio_thresh & min_dist == 1,
         ratio_lt_thresh = ifelse(is.na(ratio_lt_thresh), FALSE, ratio_lt_thresh),
         lendiff = diff_3d7 > 3 | diff_3d7 %% 3 != 0) %>% 
  ungroup() %>%
  arrange(hap_type, target, dada2_alias, reads_lt_thresh, prop_lt_thresh, ratio_lt_thresh, lendiff) %>% 
  mutate(samp_hap = factor(paste0(hap_type, target, dada2_alias, sample))) 

optim_dat_long <- optim_dat %>%
  pivot_longer(c(reads_lt_thresh, prop_lt_thresh, ratio_lt_thresh, lendiff))

hap_censor_info <- optim_dat_long %>%
  group_by(sample, dada2_alias, target, dens, hap_type, samp_hap) %>% 
  summarize(censors = list(name[value])) %>% 
  rowwise() %>% 
  mutate(censored = ifelse(length(censors) == 0, 'Not censored', 'Censored')) %>% 
  ungroup()

dada2_cens <- dada2_long %>% 
  left_join(hap_censor_info) %>% 
  filter(censored != 'Censored')

missing_info <- missing_info %>% 
  left_join(hap_censor_info)
```

Number of haplotype occurrences and haplotypes pre- and post-censoring

```{r}
(n_pre_cens <- dada2_long %>% 
  nrow())

(n_post_cens <- dada2_cens %>% 
  nrow())

n_post_cens/n_pre_cens


(n_pre_cens <- dada2_long %>% 
    select(dada2_alias) %>% 
    distinct() %>% 
  nrow())

(n_post_cens <- dada2_cens %>% 
    select(dada2_alias) %>% 
    distinct() %>% 
  nrow())

n_post_cens/n_pre_cens
```

Number of samples that returned no haplotypes post-censoring

```{r}
dada2_long %>% 
  left_join(hap_censor_info) %>% 
  group_by(sample, target) %>% 
  tally(censored != 'Censored') %>% 
  filter(n == 0) %>% 
  nrow()
```

Upset plot of how haplotypes are censored

```{r}
(fp_upset <- optim_dat_long %>%
  filter(value) %>%
  group_by(target, dens, hap_type, samp_hap) %>% 
   mutate(name = factor(case_when(name == 'reads_lt_thresh' ~ 'Read depth < threshold',
                           name == 'ratio_lt_thresh' ~ 'Read ratio < threshold',
                           name == 'lendiff' ~ 'Length different than ref',
                           name == 'prop_lt_thresh' ~ 'Read proportion < threshold'),
                       levels = c('Read depth < threshold', 'Read proportion < threshold', 
                                  'Read ratio < threshold', 'Length different than ref'))) %>% 
  summarize(censors = list(unique(name[value]))) %>% 
  rowwise() %>% 
  mutate(censored = ifelse(length(censors) == 0, 'Not censored', 'Censored')) %>% 
  ungroup() %>% 
  ggplot(aes(x=censors)) +
  facet_grid(dens~., scales = 'free', space = 'free') +
   # geom_bar(fill = 'black') +
    geom_bar(aes(fill = hap_type), color = 'white', size = 0.25) +
    scale_x_upset(order_by = 'freq') +
  scale_fill_manual(values = (c('Expected reference' = 'grey70', 'Systematic error' = 'pink', 'Random error' = 'red')),
  breaks = levels(dada2_long$hap_type),
  labels = gsub(' ', '\n', levels(dada2_long$hap_type))) +
   # scale_alpha_manual(values = rev(c(0.5, 0.75, 1)), labels = c('≥75', '1.5-75', '<1.5')) +
   labs(x = 'How censored', y = 'Number of haplotypes censored',
        fill = 'Haplotype category') +
  theme(axis.title.y = element_text(vjust = -30)))
```

Summary of censored and uncensored haplotypes

```{r}
optim_dat_long %>%
  group_by(target, dens, hap_type, samp_hap) %>% 
  summarize(censors = list(name[value])) %>% 
  rowwise() %>% 
  mutate(censored = ifelse(length(censors) == 0, 'Not censored', 'Censored')) %>% 
  ungroup() %>% 
  group_by(censored, hap_type) %>% 
  summarize(count = n()) %>% 
  group_by(hap_type) %>% 
  mutate(tot = sum(count),
         pct = round(count/tot*100))
```

Summary of how haplotypes were censored

```{r}
optim_dat_long %>% 
  ungroup() %>% 
  filter(hap_type == 'Random error') %>% 
  mutate(n_fps = n_distinct(samp_hap)) %>% 
  filter(value) %>% 
  group_by(name, n_fps) %>% 
  summarize(n = n()) %>% 
  mutate(pct = round(n/n_fps*100))
```

Comparison of haplotypes that passed and did not pass censoring

```{r}
(pass_censor <- hap_censor_info %>% 
  ggplot(aes(x = target)) +
  facet_grid(censored~., scales = 'free', space = 'free') +
  geom_bar(aes(fill = hap_type), color = 'white', size = 0.25) +
  scale_fill_manual(values = (c('grey70', 'pink', 'red'))) +
   labs(x = '', y = 'Number of haplotypes', #\nthat passed censoring', 
        fill = 'Haplotype category')) 
```

Censoring results by density

```{r}
(cens_by_dens <- optim_dat_long %>% 
               group_by(sample, target, dada2_alias) %>% 
               mutate(value = any(value)) %>% 
  filter(hap_type == 'Random error') %>%
  group_by(sample, target, dada2_alias, dens, hap_type, samp_hap, value) %>% 
  summarize() %>% 
         mutate(value = ifelse(value, 'Yes', 'No')) %>% 
  group_by(target, dens, hap_type, value) %>% 
  summarize(count = n()) %>% 
  group_by(target, dens) %>% 
  mutate(tot = sum(count),
         pct = round(count/tot*100), 
         dens = factor(gsub(' g/µL', '', dens), levels = c('<1.5', '1.5-75', '≥75'))) %>% 
  ggplot(aes(x = dens, y = count, fill = value)) +
  facet_grid(target~., scales = 'free', space = 'free') +
  geom_col(position = 'fill') +
   scale_fill_manual(values = c('grey40', 'grey80')) +
  labs(x = 'Parasite density\n(genomes/µL)', y = 'Proportion of haplotypes likely\narising from random error', fill = 'Censored') +
  coord_flip())
```

Censoring of false positive haplotypes by density

```{r}
cens_dens <- optim_dat_long %>% 
               group_by(sample, target, dada2_alias) %>% 
               mutate(value = any(value)) %>% 
  filter(hap_type == 'Random error') %>%
  group_by(sample, target, dada2_alias, dens, hap_type, samp_hap, value) %>% 
  summarize() %>% 
         mutate(value = ifelse(value, 'Yes', 'No')) %>% 
  group_by(target, dens, hap_type, value) %>% 
  summarize(count = n()) %>% 
  group_by(target, dens) %>% 
  mutate(tot = sum(count),
         pct = round(count/tot*100), 
         dens = factor(gsub(' g/µL', '', dens), levels = c('<1.5', '1.5-75', '≥75'))) %>% 
  filter(target != '*trap*') %>% 
  group_by(dens, value) %>% 
  summarize(count = sum(count)) %>% 
  pivot_wider(names_from = value, values_from = count, values_fill = 0) %>% 
  ungroup()  %>% 
  mutate(tot = (No+Yes),
         prop = No/tot)

cens_dens

cens_dens %>% 
  select(-c(dens, tot, prop)) %>% 
  fisher.test()
```


Censored true positives

```{r}
(true_cens_depth <- missing_info %>% 
  mutate(dens = factor(gsub(' g/µL', '', dens), levels = c('<1.5', '1.5-75', '≥75'))) %>% 
  filter(censored == 'Censored') %>%
  ggplot(aes(x = dens, y = ref_prop*100, col = target)) +
  geom_point(alpha = 0.5, position = position_jitter(width = 0.25, height = 0)) +
   scale_color_manual(values = marker_cols) + 
  labs(x = 'Parasite density\n(genomes/µL)', y = 'Reference haplotype percent\nof censored haplotype', col = 'Marker') +
  theme(legend.text = element_markdown()))
```

Characteristics of censored true positives

```{r}
missing_info %>% 
  mutate(dens = factor(gsub(' g/µL', '', dens), levels = c('<1.5', '1.5-75', '≥75'))) %>% 
  filter(censored == 'Censored') %>% 
  mutate(tot = n()) %>% 
  group_by(dens, tot) %>% 
  tally() %>% 
  mutate(prop = n/tot)

missing_info %>% 
  mutate(dens = factor(gsub(' g/µL', '', dens), levels = c('<1.5', '1.5-75', '≥75'))) %>% 
  filter(censored == 'Censored') %>% 
  mutate(tot = n()) %>% 
  group_by(ref_pct >= 10, tot) %>% 
  tally() %>% 
  mutate(prop = n/tot)
```


```{r}
cens1 <- plot_grid(depth_sens_spec + theme(legend.position = 'none'), 
                             prop_sens_spec + theme(legend.position = 'none'), 
          ratio_sens_spec + theme(legend.position = 'right'), 
          nrow = 1, axis = 'bt', align = 'h', rel_widths = c(1, 1, 1.55), labels = 'AUTO') 

cens2 <- plot_grid(fp_upset + theme(legend.position = c(-1, 0.5),
                                          axis.title.y = element_text(vjust = -40),
                                    plot.margin = unit(c(5.5,5.5,5.5,40), 'points')) +
                     guides(fill = guide_legend(ncol = 1)), 
                     pass_censor + theme(legend.position = 'none', 
                                         axis.text.x = element_markdown(angle = 90, hjust = 1, vjust = 0.5)),
                     cens_by_dens + theme(legend.position = 'bottom'),
                   true_cens_depth,
                     nrow = 1, labels = c('D', 'E', 'F', 'G'), axis = 'bt', align = 'vh', 
                   rel_widths = c(1, 0.5, 0.8, 0.8)
                   )

cens <- plot_grid(cens1, cens2, ncol = 1, rel_heights = c(0.8, 1))

ggsave(plot = cens, filename = 'figures/cens.png', width = 11, height = 6)
```

### Replicate variability

Percent agreement across replicates

```{r}
(rep_fig <- dada2_cens %>% 
  group_by(mix, target, dens, dada2_alias) %>% 
  tally() %>% 
  ggplot(aes(x = target, fill = as.character(n))) +
  facet_grid(dens~paste0('mix ', mix)) +
  geom_bar() +
  scale_fill_grey(start = 0.8, end = 0.4) +
  scale_y_continuous(breaks = pretty_breaks()) +
  theme(axis.text.x = element_markdown(angle = 90, hjust = 1, vjust = 0.5)) +
  labs(x = '', y = 'Number of distinct haplotypes\nfound across replicates', fill = 'Number of replicates\nin which haplotype\nwas found'))
```

```{r}
dada2_cens %>% 
  group_by(mix, target, dens, dada2_alias) %>% 
  tally() %>% 
  ungroup() %>% 
  mutate(tot = n()) %>% 
  group_by(n, tot) %>% 
  tally() %>% 
  mutate(prop = nn/tot)

dada2_cens %>% 
  group_by(mix, target, dens, dada2_alias) %>% 
  tally() %>% 
  ungroup() %>% 
  group_by(dens) %>% 
  mutate(tot = n()) %>% 
  group_by(dens, n, tot) %>% 
  tally() %>% 
  mutate(prop = nn/tot)
```

Pairwise Jaccard distance of replicates

```{r}
jac_reps <- dada2_cens %>% 
  group_by(target, dens, mix, rep) %>% 
  summarize(haps = c(dada2_alias)) %>% 
  pivot_wider(names_from = rep, values_from = haps, names_prefix = 'rep') %>% 
  ungroup() %>% 
  rowwise() %>%
  mutate(jac12 = length(intersect(rep1, rep2))/length(union(rep1, rep2)),
         jac13 = length(intersect(rep1, rep3))/length(union(rep1, rep3)),
         jac23 = length(intersect(rep3, rep2))/length(union(rep3, rep2))) %>% 
  ungroup() %>% 
  select(-c(rep1, rep2, rep3)) %>% 
  pivot_longer(c(jac12, jac13, jac23)) %>% 
  mutate(dens = factor(gsub(' g/µL', '', dens), levels = c('<1.5', '1.5-75', '≥75'))) 

(jac_fig <- jac_reps %>% 
  ggplot(aes(x = dens, y = value)) +
  geom_boxplot() +
  geom_point(aes(col = target), position = position_jitter(width = 0.3, height = 0), alpha = 0.4) +
  scale_color_manual(values = marker_cols_light) +
  labs(y = 'Pairwise Jaccard distance\nbetween replicates', x = 'Parasite density\n(genomes/µL)', col = 'Marker') +
  theme(legend.text = element_markdown()))
```

```{r}
jac_reps %>% 
  summarize(median_jac = median(value, na.rm = TRUE),
            iqr_jac = IQR(value, na.rm = TRUE))

jac_reps %>% 
  group_by(dens) %>% 
  summarize(median_jac = median(value, na.rm = TRUE),
            iqr_jac = IQR(value, na.rm = TRUE))
```


```{r}
rep_comp_fig <- plot_grid(rep_fig, jac_fig, nrow = 1, labels = 'AUTO', align = 'hv', rel_widths = c(1.5, 1), axis = 'trbl')
ggsave(plot = rep_comp_fig, filename = 'figures/reps.png', width = 10, height = 3)
```



### Missing haplotypes

Number of expected haplotypes that were missing

```{r}
missing_info %>% 
  mutate(tot = n()) %>% 
  group_by(found, tot) %>% 
  tally() %>% 
  mutate(prop = n/tot)
```
Found and missing haplotypes by reference percent

```{r}
pct_missing <- missing_info %>% 
  group_by(target, dens) %>% 
  summarize(n_missing = sum(prop_reads > 0),
            n_found = sum(prop_reads == 0),
            tot = n()) %>% 
  mutate(missing = round(100-n_missing/tot*100),
         found = round(100-n_found/tot*100)) %>% 
  select(target, dens, missing, found) %>% 
  pivot_longer(c(missing, found)) %>% 
  mutate(found = ifelse(name == 'found', 'Yes', 'No'))

(missing_hist <- missing_info %>% 
   mutate(found = case_when(found == 'No' ~ 'No',
                            found == 'Yes' & censored == 'Censored' ~ 'Yes (censored)',
                            found == 'Yes' & censored == 'Not censored' ~ 'Yes (not censored)')) %>%
   mutate(dens = factor(dens, levels = rev(levels(dens)))) %>% 
  ggplot(aes(x = ref_pct, fill = found, col = found)) +
  facet_grid(target~dens) +
  geom_histogram() +
  geom_text(data = pct_missing %>% filter(found == 'No'), x = 90, y = 35, #y = 27,
            aes(label = paste0(value, '%')), check_overlap = TRUE, size = 3) +
  scale_fill_grey(start = 0.7, end = 0.3) +
  scale_color_grey(start = 0.7, end = 0.3) +
   guides(color = 'none') +
  labs(x = 'Reference haplotype percent', y = 'Number of haplotypes', fill = 'Haplotype\nobserved'))
```

Proportion missing by read depth

```{r}
# proportion found vs. read depth
(missing_read_depth <- missing_info %>% 
   mutate(found = case_when(found == 'No' ~ 'No',
                            found == 'Yes' & censored == 'Censored' ~ 'Yes (censored)',
                            found == 'Yes' & censored == 'Not censored' ~ 'Yes (not censored)')) %>%
  group_by(sample, target, dens, mix, n_reads_sample) %>% 
  summarize(prop_missing = 1 - sum(found == 'Yes (not censored)')/n()) %>% 
   mutate(dens = factor(gsub(' g/µL', '', dens), levels = c('<1.5', '1.5-75', '≥75'))) %>% 
  ggplot(aes(x = n_reads_sample, y = prop_missing)) +
  geom_point(aes(shape = dens, col = target), alpha = 0.5, size = 3) +
   stat_cor(aes(label = ..r.label..), method = 'spearman', label.y = 0.95, cor.coef.name = 'rho', r.accuracy = 0.01) +
  scale_color_manual(values = marker_cols) +
  scale_x_continuous(breaks = pretty_breaks()) +
  labs(x = 'Sample read depth', y = 'Proportion of missing haplotypes',
       col = 'Marker', shape = 'Parasite density\n(genomes/µL)') +
   theme(legend.text = element_markdown()))
```

Correlation between reference and read proportion

```{r}
(missing_prop <- missing_info %>% 
   mutate(dens = factor(dens, levels = rev(levels(dens)))) %>% 
  ggplot(aes(x = ref_pct, y = n_reads/n_reads_sample*100, col = target, group = '')) +
  facet_grid(~dens) +
  geom_point(alpha = 0.2) +
   stat_poly_eq() +
  scale_color_manual(values = marker_cols) +
  labs(x = 'Reference haplotype percent\n(expected percent of reads)', y = 'Observed percent of reads', col = 'Marker') +
   theme(legend.text = element_markdown()))
```

Replicate missingness by reference proportion

```{r}
(missing_rep <- missing_info %>% 
  group_by(target, dens, mix, ref_pct, ref_name_all) %>% 
  summarize(in_n_rep = sum(n_reads > 0)) %>% 
  mutate(ref_pct_bin = factor(case_when(ref_pct < 10 ~ '<10',
                                      ref_pct >= 10 ~ '≥10'),
                                   levels = c('<10', '≥10')),
         dens = factor(dens, levels = rev(levels(dens)))) %>% 
  group_by(dens, ref_pct_bin, in_n_rep) %>%
  mutate(count = n()) %>% 
  group_by(dens, ref_pct_bin) %>%
  mutate(tot = n(),
         pct = count/tot*100) %>% 
  ggplot(aes(x = ref_pct_bin, y = as.character(in_n_rep))) +
  facet_grid(~dens) +
  geom_tile(aes(fill = pct)) +
  geom_jitter(aes(col = target), alpha = 0.5, show.legend = FALSE) +
  scale_fill_gradient(low = 'white', high = 'grey70') +
  scale_color_manual(values = marker_cols) +
  labs(x = 'Strain percent in mixture', 
       y = 'Number of replicates in which\nhaplotype was found',
       col = 'Marker', 
       fill = 'Percent of\nhaplotypes'))
```

Number of replicates in which haplotype was found

```{r}
missing_info %>% 
  group_by(target, dens, mix, ref_pct, ref_name_all) %>% 
  summarize(in_n_rep = sum(n_reads > 0)) %>% 
  mutate(ref_pct_bin = factor(case_when(ref_pct < 10 ~ '<10',
                                      ref_pct >= 10 ~ '≥10'),
                                   levels = c('<10', '≥10'))) %>% 
  group_by(dens, in_n_rep) %>%
  summarize(count = n()) %>% 
  group_by(dens) %>%
  mutate(tot = sum(count),
         pct = round(count/tot*100))
```


```{r}
missing_fig <- plot_grid(missing_hist + theme(legend.position = 'left'), 
          missing_read_depth + 
            scale_x_continuous(breaks = c(0, 40000, 80000)),  
          missing_prop + theme(legend.position = 'left'), 
          missing_rep + guides(color = 'none'),
          labels = 'AUTO', rel_heights = c(1, 0.9), axis = 'trbl', align = 'h')
ggsave(plot = missing_fig, filename = 'figures/missing.png', width = 10, height = 7)
```


```{r}
missing_lm_target <- glmer(missing ~ target + (1|mix_dens), 
    data = missing_info, family=binomial(link='logit'))

missing_lm_ref_prop <- glmer(missing ~ ref_pct + (1|mix_dens), 
    data = missing_info, family=binomial(link='logit'))

missing_lm_reads <- glmer(missing ~ reads_10000 + (1|mix_dens), 
    data = missing_info, family=binomial(link='logit'))

missing_lm_dens <- glmer(missing ~ dens + (1|mix_dens), 
    data = missing_info, family=binomial(link='logit'))

missing_lm_moi <- glmer(missing ~ expected_moi + (1|mix_dens), 
    data = missing_info, family=binomial(link='logit'))

missing_lm_all <- glmer(missing ~ ref_pct + target + dens + reads_10000 + expected_moi + (1|mix_dens),
    data = missing_info, family=binomial(link='logit'))
```

Missingness risk factors

```{r}
lm_biv_summary <- lapply(list(missing_lm_ref_prop, missing_lm_target, missing_lm_dens, missing_lm_reads, missing_lm_moi), function(x){
  x %>% tidy(exponentiate = TRUE, conf.int = TRUE) 
}) %>% bind_rows()

lm_multi_summary <- missing_lm_all %>% tidy(conf.int = TRUE, exponentiate = TRUE) 

lm_biv_summary %>%
  filter(term != '(Intercept)' & term != 'sd__(Intercept)') %>%
  mutate(bivariate = paste0(round(estimate, 2),
                            ' (', round(conf.low, 2), '-', round(conf.high, 2), '); p=',
                            signif(p.value, 2)),
         feature = c('ref_prop', rep('target', 4), rep('density', 2), 'reads', 'moi')) %>%
  select(feature, term, bivariate) %>%
  left_join(lm_multi_summary %>% 
  filter(term != '(Intercept)' & term != 'sd__(Intercept)') %>% 
  mutate(multivariate = paste0(round(estimate, 2), 
                            ' (', round(conf.low, 2), '-', round(conf.high, 2), '); p=',
                            signif(p.value, 1)),
         feature = c('ref_prop', rep('target', 4), rep('density', 2), 'reads', 'moi')) %>% 
  select(feature, term, multivariate)) #%>% View()

```


### MOI & clinical samples

Observed vs expected MOI

```{r}
(mix_moi_fig <- dada2_cens %>% 
  group_by(sample, target, mix, dens, expected_moi) %>% 
  summarize(est_moi = n()) %>% 
  mutate(obs_min_exp_moi = est_moi - expected_moi) %>% 
  ggplot(aes(x = obs_min_exp_moi, fill = target)) +
  facet_grid(dens~.) +
  geom_bar(fill = 'white') +
  geom_bar() +
  scale_fill_manual(values = marker_cols_light) +
  labs(x = 'Observed MOI - expected MOI', y = 'Number of samples', fill = 'Marker') +
  theme(legend.text = element_markdown()))
```


```{r}
dada2_cens %>% 
  mutate(dens = factor(gsub(' g/µL', '', dens), levels = c('<1.5', '1.5-75', '≥75'))) %>% 
  group_by(sample, target, mix, dens, expected_moi) %>% 
  summarize(est_moi = n()) %>% 
  mutate(obs_min_exp_moi = est_moi - expected_moi,
         obs_min_exp_moi_cat = case_when(obs_min_exp_moi == 0 ~ 'Same as expected',
                                         obs_min_exp_moi > 0 ~ 'Higher than expected',
                                         obs_min_exp_moi < 0 ~ 'Lower than expected')) %>% 
  ungroup() %>% 
  mutate(tot = n()) %>% 
  group_by(obs_min_exp_moi_cat, tot) %>% 
  tally() %>% 
  mutate(pct = round(n/tot*100))
```

```{r}
dada2_cens %>% 
  mutate(dens = factor(gsub(' g/µL', '', dens), levels = c('<1.5', '1.5-75', '≥75'))) %>% 
  group_by(sample, target, mix, dens, expected_moi) %>% 
  summarize(est_moi = n()) %>% 
  mutate(obs_min_exp_moi = est_moi - expected_moi) %>% 
  ungroup() %>% 
  summarize(median_low = median(obs_min_exp_moi[dens == '<1.5']),
            median_mid = median(obs_min_exp_moi[dens == '1.5-75']),
            median_high = median(obs_min_exp_moi[dens == '≥75']),
            wilcox_p = wilcox.test(obs_min_exp_moi[dens == '<1.5'], 
                                   obs_min_exp_moi[dens != '<1.5'])$p.value)
```

Clinical samples

```{r}
dada2_long_clinsamps <- dada2_long_clinsamps %>% 
  group_by(sample_id, target) %>% 
  mutate(read_thresh = optim_depth$optimal_thresholds %>% 
           filter(cost == 'FN 2x cost of FP' & metric == 'threshold') %>% 
           pull(threshold),
         prop_thresh = optim_prop$optimal_thresholds %>% 
           filter(cost == 'FN 2x cost of FP' & metric == 'threshold') %>% 
           pull(threshold),
         ratio_thresh = optim_ratio$optimal_thresholds %>% 
           filter(cost == 'FN 2x cost of FP' & metric == 'threshold') %>% 
           pull(threshold),
         reads_lt_thresh = n_reads < read_thresh,
         prop_lt_thresh = n_reads/sum(n_reads) < prop_thresh,
         ratio_lt_thresh = read_prop <= ratio_thresh & min_dist == 1,
         ratio_lt_thresh = ifelse(is.na(ratio_lt_thresh), FALSE, ratio_lt_thresh)) %>% 
  ungroup() %>% 
        mutate(diff_len = diff_3d7 > 3 | diff_3d7 %% 3 != 0)
```

```{r}
dada2_long_clinsamps %>%
  mutate(censored = reads_lt_thresh | prop_lt_thresh | ratio_lt_thresh | diff_len) %>% 
  mutate(tot = n()) %>%
  group_by(censored, tot) %>% 
  tally() %>% 
  mutate(pct = round(n/tot*100))
```


```{r}
est_moi <- dada2_long_clinsamps %>% 
  filter(!reads_lt_thresh & !prop_lt_thresh & !ratio_lt_thresh & !diff_len) %>% 
  group_by(sample_id, target) %>% 
  summarize(est_moi = n())

pct_mode <- est_moi %>% 
left_join(est_moi %>% 
  group_by(sample_id, est_moi) %>% 
  tally() %>% 
  group_by(sample_id) %>% 
  slice_max(n, with_ties = FALSE)) %>% 
  mutate(mode = ifelse(is.na(n), 'no', 'yes')) %>% 
  select(sample_id, mode) %>% 
  group_by(sample_id) %>% 
  summarize(pct_mode = mean(mode == 'yes')*100) %>% 
  ungroup() %>% 
  arrange(-pct_mode) %>% 
  mutate(sample_id = fct_inorder(sample_id))

est_moi <- est_moi %>% 
  mutate(sample_id = factor(sample_id, levels = levels(pct_mode$sample_id)))


(clin_fig <- plot_grid(est_moi %>% 
  ggplot(aes(y = sample_id, x = target, fill = as.character(est_moi))) +
  geom_tile(col = 'white') +
  scale_fill_grey(start = 0.9, end = 0.1) +
  labs(x = 'Marker', y = 'Clinical sample', fill = 'Estimated\nMOI') +
  theme(legend.position = 'left',
        axis.text.y = element_blank(),
        axis.ticks = element_blank()),

pct_mode %>% 
  ggplot(aes(y = sample_id, fill = fct_rev(paste0(pct_mode, '%')))) +
  geom_bar() +
  scale_fill_viridis_d(option = 'mako') +
  theme_void() +
  theme(legend.position = 'right') +
  labs(fill = 'Percent\nconcordance'), align = 'h', rel_widths = c(2.5, 1)))
```

```{r}
moi_fig <- plot_grid(mix_moi_fig + theme(legend.position = 'left'), clin_fig, labels = 'AUTO', rel_widths = c(1, 1.25))
ggsave(plot = moi_fig, filename = 'figures/moi.png', width = 8, height = 3)
```


